### DISCUSSION TITLE
Atoms are updated but not rendered

### DISCUSSION THREAD

**Discussion by hikagami0210** [pmndrs/jotai#2401] [Q&A]
I am not good at English, so I use translations.

Jotai is used to fetch from api.
In doing so, the fetched data is set to PrimitiveAtom using WritableAtom, but components that reference PrimitiveAtom with useAtom (or sueAtomValue) are not rendered. If the screen is re-rendered by other factors (e.g. buttons), it renders correctly. What could be the cause?

<details><summary>my code</summary>

`atoms`
```ts
export const eventObject = atom<{
    [key: string]: any[];
}>({});

const objectFormat = (newData: any, preData: any) => {
    const fetchEventObject = preData;
    Object.entries(newData).forEach(([date, events]) => {
       // processing
    })
    return fetchEventObject;
};

export const writableAtom = atom(null, async (get, set) => {
    const url = "url";
    await axios.get(url).then((response) => {
        console.log(`------axios.get------`);

        const preData = get(topPageEventListObject);
        const newData = response.data;
        const fetchEventObject = objectFormat(newData, preData);

        set(topPageEventListObject, fetchEventObject); 
        console.log(`------set fetchEventObject------`);
    });
});
```

`components`
```tsx
const EventList: React.FC<TopPageEventListProps> = () => {
    const eventList = useAtomValue(eventObject);
    const fetch = useSetAtom(writableAtom);

    useEffect(() => {
            fetch();
    }, [eventList, fetch]);

    console.log("EventList render");
    console.log(eventList);

    return (
        <>
            {Object.entries(eventList).map(([date, events]) => {
                return (
                    <React.Fragment key={date}>
                        // div...
                    </React.Fragment>
                );
            })}
        </>
    );
};
```

`console`
```txt
// window reloading
EventList render
{} // Correct Objects, But the display is different from the later one

EventList render
{} // Correct Objects

------axios.get------
------set fetchEventObject------
------axios.get------
------set fetchEventObject------

// button click
// EventList will be displayed correctly at this time

EventList render
{2025-08-21: Array(7), 2032-01-01: Array(3)}
```
</p>
</details>

**Comment by dai-shi** [comment]
You shouldn't mutate atom values. Assuming your event data are immutable, here's an easy fix:
```diff
- const fetchEventObject = preData;
+ const fetchEventObject = { ...preData };
```

**Reply by hikagami0210** [reply]
@dai-shi 

Thank you.
I changed the code, but it did not solve the problem.

**Reply by dai-shi** [reply]
How about this?

```diff
- const fetchEventObject = objectFormat(newData, preData);
+ const fetchEventObject = newData;
```

**Reply by hikagami0210** [reply]
@dai-shi 

Oh! It was displayed correctly.

However, since this application has the function of infinite loading, it is necessary to add new data to old data while also processing to avoid duplication.

That is the objectFormat method, but is this method wrong for using jotai...

**Reply by dai-shi** [reply]
You have to make `objectFormat` function so that it doesn't mutate objects.

You can test with deepFreeze. (like https://www.npmjs.com/package/deep-freeze-strict)

```js
const objectFormat = (newData: any, preData: any) => {
    deepFreeze(preData); // only for debugging
    const fetchEventObject = { ...preData };
    Object.entries(newData).forEach(([date, events]) => {
       // processing, do not mutate objects!
    })
    return fetchEventObject;
};
```

**Reply by dai-shi** [reply]
> it is necessary to add new data to old data while also processing to avoid duplication.

Oh, I see. But, it doesn't work with React state, I suppose. It's not Jotai's limitation.

**Comment by dai-shi** [comment]
Actually, we have `freezeAtom`.

https://github.com/pmndrs/jotai/blob/4d8b6f5d970d16aeb107c37ae2581199424aa476/docs/guides/debugging.mdx#L132

So, this will help debugging:

```ts
export const eventObject = freezeAtom(atom<{
    [key: string]: any[];
}>({}));
```

**Comment by hikagami0210** [comment]
@dai-shi 


I was trying different things, I have confirmed that the following approach works correctly.

```ts
const objectFormat = (newData: any, preData: any) => {
    const fetchEventObject = { ...JSON.parse(JSON.stringify(preData)) };
```

It is thanks to you that I realized that the behavior of the object was the problem.
Thank you very much.

Sorry for asking a question that is not really related to jotai.

**Reply by dai-shi** [reply]
It works too. Though, deep cloning isn't very efficient. Ideally, you should only copy object if necessary. You might want to consider using immer, or something.

**Reply by hikagami0210** [reply]
The final result looked like this.

Thank you very much for teaching me how to do this!

```ts
const objectFormat = (newData: any, preData: any) => {
    const draft = createDraft(preData);
    Object.entries(newData).forEach(([date, events]) => {
     // processing
    });

    const nextState = finishDraft(draft);
    return nextState;
};
```
